src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs

```diff
@@ -2519,6 +2519,11 @@ internal void ReadSni(TaskCompletionSource<object> completion)
                     Debug.Assert(IsValidPacket(readPacket), "ReadNetworkPacket should not have been null on this async operation!");
                     // Evaluate this condition for MANAGED_SNI. This may not be needed because the network call is happening Async and only the callback can receive a success.
                     ReadAsyncCallback(IntPtr.Zero, readPacket, 0);
+
+                    if (!IsPacketEmpty(readPacket))
+                    {
+                        ReleasePacket(readPacket);
+                    }
```

Comment:
This looks redundant because of the check after this block of code for ManagedSNI before doing a packet release but it isn't. I tried putting just one release in place but any other combination causes a leak either in managed or native.

In Managed mode we have to release the packet only if there was a synchronous result. 
In Native mode we have to release the packet in if we received one.